numba/jitclass/base.py

```diff
@@ -0,0 +1,362 @@
+        assert not argspec.defaults, 'defaults not supported'
+        args = ', '.join(argspec.args[1:])
+        ctor_source = _ctor_template.format(args=args)
+        glbls = {"__numba_cls_": cls}
+        exec_(ctor_source, glbls)
+        ctor = glbls['ctor']
+        cls._ctor = njit(ctor)
+
+    def __instancecheck__(cls, instance):
+        if isinstance(instance, boxing.Box):
+            return instance._numba_type_.class_type is cls.class_type
+        return False
+
+    def __call__(cls, *args, **kwargs):
+        return cls._ctor(*args, **kwargs)
+
+
+##############################################################################
+# Registration utils
+
+def register_class_type(cls, spec, class_ctor, builder):
+    """
+    Internal function to create a jitclass.
+
+    Args
+    ----
+    cls: the original class object (used as the prototype)
+    spec: the structural specification contains the field types.
+    class_ctor: the numba type to represent the jitclass
+    builder: the internal jitclass builder
+    """
+    # Normalize spec
+    if isinstance(spec, Sequence):
+        spec = OrderedDict(spec)
+
+    # Copy methods from base classes
+    clsdct = {}
+    for basecls in reversed(inspect.getmro(cls)):
+        clsdct.update(basecls.__dict__)
+
+    methods = dict((k, v) for k, v in clsdct.items()
+                   if isinstance(v, pytypes.FunctionType))
+    others = dict((k, v) for k, v in clsdct.items() if k not in methods)
+    docstring = others.pop('__doc__', "")
+    _drop_ignored_attrs(others)
+    if others:
+        msg = "class members are not yet supported: {0}"
+        members = ', '.join(others.keys())
+        raise TypeError(msg.format(members))
+
+    jitmethods = {}
+    for k, v in methods.items():
+        jitmethods[k] = njit(v)
+
+    class_type = class_ctor(cls, spec, jitmethods)
+    cls = JitClassType(cls.__name__, (cls,), dict(class_type=class_type,
+                                                  __doc__=docstring))
+
+    # Register resolution of the class object
+    typer = CPUTarget.typing_context
+    typer.insert_global(cls, class_type)
+
+    # Register class
+    backend = CPUTarget.target_context
+    builder(class_type, methods, typer, backend).register()
+
+    return cls
+
+
+def _drop_ignored_attrs(dct):
+    # ignore anything defined by object
+    drop = set(['__weakref__',
+                '__module__',
+                '__dict__'])
+    for k, v in dct.items():
+        if isinstance(v, (pytypes.BuiltinFunctionType,
+                          pytypes.BuiltinMethodType)):
+            drop.add(k)
+        elif getattr(v, '__objclass__', None) is object:
+            drop.add(k)
+
+    for k in drop:
+        del dct[k]
+
+
+class ClassBuilder(object):
+    """
+    A jitclass builder for mutable jitclasses.
+    """
+    instance_type_class = types.ClassInstanceType
+    per_backend_registry = {}
+
+    def __init__(self, class_type, methods, typer, backend):
```

Comment:
Just a nit, but I would rather keep calling "backend" "targetctx" or something like that. Otherwise it gets confusing.